students/MicahBraun/Lesson 05/mailroom_3.py

```diff
@@ -0,0 +1,297 @@
+#  to the architecture of the rest of the surrounding code, I left it as-is. **
+#
+# While I would like to add comprehensions to my program to simplify things or make it more succinct, I did
+# not find any workable situations for me to apply the examples I had with the way I had structured things
+# (short of re-writing the entire thing just to add comprehensions...)
+# ----------------------------------------------------------------------------------------------------------------------
+
+import time     # time to supply today's date for letters
+import os       # for default working directory
+import locale   # for currency formatting
+import sys      # for error handling
+#  --------------------------------------------------- DATA ------------------------------------------------------------
+
+locale.setlocale(locale.LC_ALL, '')
+today = time.strftime("%m/%d/%Y")
+
+spacing = '- ' * 56  # formatting for DONOR header
+donor_dict = {'Columns': ['NAMES', 'DONATION AMOUNT', 'NUMBER OF GIFTS', 'AVG. GIFTS'],
+              'Rudolph Soares': ['Rudolph Soares', 1335, 3, 445],  # Donor info
+              'Josef Mistretta': ['Josef Mistretta', 7500, 4, 1875],
+              'Joye Agostini': ['Joye Agostini', 600, 2, 300],
+              'Rachelle Levan': ['Rachelle Levan', 12100, 5, 2420],
+              'Vena Ussery': ['Vena Ussery', 4000, 8, 500],
+              'Efrain Lager': ['Efrain Lager', 795, 9, 88.34],
+              'Mee Heine': ['Mee Heine', 4600, 4, 1150],
+              'Tanya Essex': ['Tanya Essex', 75000, 2, 37500]}
+
+
+#  ------------------------------------------------- PROCESSING --------------------------------------------------------
+
+def menu():
+    """Display menu of options to toggle between to perform separate actions"""
+    header = "-" * 110  # formatting for header
+    title = ' ' * 48 + "MAIL ROOM MENU"  # title display
+    directions = "Select one of the three options:"  # directions to user for menu
+    optA = "[A] Send a Thank you"  # option(s) for menu
+    optB = "[B] Create a Report"
+    optC = "[C] Create Letters for All Donors"
+    optD = "[D] Quit"
+    whole_menu_header = header + '\n' + title + '\n' + header  # header strung together as one uni
+
+    while True:
+        print(whole_menu_header)  # display header
+        print(directions.center(110, ' '), '\n')  # center directions with header display
+        print(optA + optB.center(37, ' ') + optC.center(36, ' ') + optD.center(25, ' ') + '\n')
+
+        switch_function = {
+            'A': thankyou,  # dictionary holding menu options
+            'B': createreport,
+            'C': writeletters,
+            'D': quit_program,
+        }
+        try:
+            usr_sel = (input("Menu Selection: ")).upper()  # user input for menu
+            switch_function.get(usr_sel)()  # get user entry from dict
+
+        except Exception as e:
+            print("Invalid Entry! Only enter A, B, C, or D. \n", e)  # display error on any other entry
```

Comment:
It's almost always better to hone in on a specific exception rather than catching the base `Exception` object. In this case I would recommend using bracket syntax for the dict look up e.g. `switch_function[user_sel]()` rather than get, so you can catch `KeyError` rather than `Exception` on bad user input